commands/sandbox.go

```diff
@@ -268,12 +250,135 @@ func (c *CmdConfig) PrintSandboxTextOutput(output do.SandboxOutput) error {
+	}
+	return nil
+}
+
+// Working subroutine for 'install' and 'upgrade'
+func doSandboxInstall(sandboxDir string, upgrading bool) error {
+	// Make a temporary directory for use during the install
+	tmp, err := ioutil.TempDir("", "doctl-sandbox")
+	if err != nil {
+		return err
+	}
+	// Download the nodejs tarball for this os and architecture
+	goos := runtime.GOOS
+	arch := runtime.GOARCH
+	if arch == "amd64" {
+		arch = "x64"
+	} else if arch == "windows" {
+		arch = "win"
+	}
+	var nodeFileName string
+	var nodeDir string
+	if upgrading && canReuseNode(sandboxDir) {
+		err = copyExistingNode(sandboxDir, tmp)
+		if err != nil {
+			return err
+		}
+	} else {
+		nodeDir = fmt.Sprintf("node-%s-%s-%s", NODE_VERSION, goos, arch)
+		URL := fmt.Sprintf("https://nodejs.org/dist/%s/%s.tar.gz", NODE_VERSION, nodeDir)
+		nodeFileName = filepath.Join(tmp, "node-install.tar.gz")
+		fmt.Print("Downloading...")
+		err = download(URL, nodeFileName)
+		if err != nil {
+			return err
+		}
+	}
+	// Download the fat tarball with the nim CLI, deployer, and sandbox bridge
+	// TODO do these need to be arch-specific?  Currently assuming not.
+	URL := fmt.Sprintf("https://do-serverless-tools.nyc3.digitaloceanspaces.com/doctl-sandbox-%s.tar.gz", MIN_SANDBOX_VERSION)
+	sandboxFileName := filepath.Join(tmp, "doctl-sandbox.tar.gz")
+	err = download(URL, sandboxFileName)
+	if err != nil {
+		return err
+	}
+	// Exec tar binary twice to unpack the two tarballs into the tmp directory
+	// TODO eliminate use of separate tar binary so we can support windows install with pure go code
+	fmt.Print("Unpacking...")
+	cmd := exec.Command("tar", "-C", tmp, "-xzf", sandboxFileName)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Printf("%s", output)
+		return err
+	}
+	if nodeFileName != "" {
+		cmd := exec.Command("tar", "-C", tmp, "-xzf", nodeFileName)
+		output, err := cmd.CombinedOutput()
+		if err != nil {
+			fmt.Printf("%s", output)
+			return err
+		}
+	}
+	// Move artifacts to final location
+	fmt.Print("Installing...")
+	srcPath := filepath.Join(tmp, "sandbox")
+	if upgrading {
+		// Preserve credentials by copying them from target (which will be replaced) to source.
+		credPath := filepath.Join(sandboxDir, ".nimbella")
+		relocPath := filepath.Join(srcPath, ".nimbella")
+		err = os.Rename(credPath, relocPath)
+		if err != nil {
+			return err
+		}
+		// Remove former sandboxDir before moving in the new one
+		err = os.RemoveAll(sandboxDir)
+	}
+	err = os.Rename(srcPath, sandboxDir)
+	if err != nil {
+		return err
+	}
+	if nodeFileName != "" {
+		srcPath = filepath.Join(tmp, nodeDir, "bin", "node")
+		destPath := filepath.Join(sandboxDir, "node")
+		err = os.Rename(srcPath, destPath)
+		if err != nil {
+			return err
+		}
+	}
+	fmt.Println("\nDone")
+	return nil
+}
+
+// Check whether the installed node is at the appropriate version
+// Avoids downloading node again during an upgrade
+func canReuseNode(sandboxDir string) bool {
+	return false
```

Comment:
Oops.   I forgot to implement this.  It should capture the output of `node --version` and compare it to the node version in the constant.